steering_controllers/src/steering_controller_implementations.cpp

```diff
@@ -0,0 +1,174 @@
+  const double wheel_seperation = params_.wheel_separation_multiplier * params_.wheel_separation;
+  const double wheelbase = params_.wheelbase_multiplier * params_.wheelbase;
+  odometry_.set_wheel_params(wheel_radius, wheel_seperation, wheelbase);
+  odometry_.set_velocity_rolling_window_size(params_.velocity_rolling_window_size);
+
+  // TODO: enable position/velocity configure
+  const size_t nr_state_itfs = 4;
+  const size_t nr_cmd_itfs = 4;
+  const size_t nr_ref_itfs = 2;
+
+  set_interface_numbers(nr_state_itfs, nr_cmd_itfs, nr_ref_itfs);
+
+  RCLCPP_INFO(get_node()->get_logger(), "ackermann odom configure successful");
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+}  // namespace ackermann_steering_controller
+
+namespace bicycle_steering_controller
+{
+BicycleSteeringController::BicycleSteeringController() : steering_controllers::SteeringControllers()
+{
+}
+
+controller_interface::CallbackReturn BicycleSteeringController::configure_odometry()
+{
+  params_ = param_listener_->get_params();
+
+  const double wheel_seperation = params_.wheel_separation_multiplier * params_.wheel_separation;
+  const double wheel_radius = params_.wheel_radius_multiplier * params_.wheel_radius;
+  odometry_.set_wheel_params(wheel_radius, wheel_seperation);
+  odometry_.set_velocity_rolling_window_size(params_.velocity_rolling_window_size);
+
+  // TODO: enable position/velocity configure
+  const size_t nr_state_itfs = 2;
+  const size_t nr_cmd_itfs = 2;
+  const size_t nr_ref_itfs = 2;
+
+  set_interface_numbers(nr_state_itfs, nr_cmd_itfs, nr_ref_itfs);
+
+  RCLCPP_INFO(get_node()->get_logger(), "bicycle odom configure successful");
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+bool BicycleSteeringController::update_odometry(const rclcpp::Duration & period)
+{
+  if (params_.open_loop)
+  {
+    odometry_.update_open_loop(last_linear_velocity_, last_angular_velocity_, period.seconds());
+  }
+  else
+  {
+    const double rear_wheel_value = state_interfaces_[0].get_value();
+    const double steer_position = state_interfaces_[1].get_value() * params_.steer_pos_multiplier;
+    if (!std::isnan(rear_wheel_value) && !std::isnan(steer_position))
+    {
+      if (params_.position_feedback)
+      {
+        // Estimate linear and angular velocity using joint information
+        odometry_.update_from_position(rear_wheel_value, steer_position, period.seconds());
+      }
+      else
+      {
+        // Estimate linear and angular velocity using joint information
+        odometry_.update_from_velocity(rear_wheel_value, steer_position, period.seconds());
+      }
+    }
+  }
+  return true;
+}
+}  // namespace bicycle_steering_controller
+
+namespace tricycle_steering_controller
+{
+TricycleSteeringController::TricycleSteeringController()
+: steering_controllers::SteeringControllers()
+{
+}
+
+controller_interface::CallbackReturn TricycleSteeringController::configure_odometry()
+{
+  params_ = param_listener_->get_params();
```

Comment:
```suggestion
  For example use in AckermannController
```
  ackermann_param_listener = std::make_shared<ackermann_controllers::ParamListener>(get_node());
  ackermann_params = ackermann_param_listener->get_params();
    }
```
```